Implement `coalesce` to more closely match Terraform (#78)
pulumi/pulumi-std: CHANGELOG_PENDING.md

### Improvements

- Use latest patch version of go in CI.

> - Change coalesce to handle nil values and convert values to more general types.

Reviewer: ```suggestion
- Handle nil values in coalesce and try to convert values to the most common type.
```

Reviewer 2: The goal is to try and match the TF behaviour as closely as possible (we want to target this function when `convert`ing), so I think we have to be a bit cleverer than that. That said, there are still some cases we don't match, because TF's `coalesce` is weird (broken) in some cases:

```
$ tofu console
> coalesce(null, false, true)
false
> coalesce(null, false, "true")
"false"

# Wait what
> coalesce(null, false, 0)
false
> coalesce(null, 0, false)
0

# Seriously
> type(coalesce(null, 0, false))
number
> type(coalesce(null, false, 0))
bool
```

I'm going to update the PR comment to make this clearer though.

Reviewer: It's like Javascript!  The PR makes sense to me then since the goal is to try to match terraform bug for bug (even if some of it doesn't really make sense)
